electronics_lib/Passives.py

```diff
@@ -314,6 +297,180 @@ def derated_capacitance(row: Dict[str, Any]) -> Tuple[float, float]:
+      max=4.7e-5,
+      derate=0,
+      vc_pairs={100: 4.7e-6, 50:   1e-5,           16: 2.2e-5, 10: 4.7e-5},
+    ),
+    SmtCeramicCapacitorGenericPackageSpecs(
+      name='Capacitor_SMD:C_1812_4532Metric',
+      max=1e-4,
+      derate=0,
+      vc_pairs={100: 2.2e-6, 50:   1e-6, 25: 1e-5, },
+    ),
+  ]
+
+  def select_capacitor_no_prod_table(self, capacitance: RangeVal, voltage: RangeVal,
+                                     footprint_spec: str, derating_coeff: float) -> None:
+    """
+    Selects a generic capacitor without using product tables
+
+    :param capacitance: user-specified (derated) capacitance
+    :param voltage: user-specified voltage
+    :param single_nominal_capacitance: used when no single cap with requested capacitance, must generate multiple parallel caps,
+                                       actually refers to max capacitance for a given part
+    :param footprint_spec: user-specified package footprint
+    :param derating_coeff: user-specified derating coefficient, if used then footprint_spec must be specified
+    """
+
+    def select_package(nominal_capacitance: RangeVal, voltage: RangeVal) -> Optional[str]:
+
+      if footprint_spec == "":
+        package_options = self.PACKAGE_SPECS
+      else:
+        package_options = [spec for spec in self.PACKAGE_SPECS if spec.name == footprint_spec]
+
+      for package in package_options:
+        if package.max >= nominal_capacitance[1]:
+            for package_max_voltage, package_max_capacitance in package.vc_pairs.items():
+              if package_max_voltage >= voltage[1] and package_max_capacitance >= nominal_capacitance[1]:
+                return package.name
+      return None
+
+    nominal_capacitance = (capacitance[0] / derating_coeff, capacitance[1] / derating_coeff)
+
+    num_caps = math.ceil(nominal_capacitance[0] / self.SINGLE_CAP_MAX)
+    if num_caps > 1:
+      assert num_caps * self.SINGLE_CAP_MAX < nominal_capacitance[1], "can't generate parallel caps within max capacitance limit"
+
+      self.assign(self.selected_nominal_capacitance, (
+        num_caps * nominal_capacitance[0],
+        num_caps * nominal_capacitance[1],
+      ))
+
+      if footprint_spec == "":
+        split_package = self.MAX_CAP_PACKAGE
+      else:
+        split_package = footprint_spec
+
+      cap_model = DummyCapacitor(capacitance=(self.SINGLE_CAP_MAX, self.SINGLE_CAP_MAX),
+                                      voltage=self.voltage, footprint_spec=split_package)
+      self.c = ElementDict[DummyCapacitor]()
+      for i in range(num_caps):
+        self.c[i] = self.Block(cap_model)
+        self.connect(self.c[i].pos, self.pos)
+        self.connect(self.c[i].neg, self.neg)
+    else:
+      valid_footprint_spec = select_package(nominal_capacitance, voltage)
+      assert valid_footprint_spec is not None, "cannot generate a valid footprint spec"
+      value = choose_preferred_number(nominal_capacitance, 0, self.E24_SERIES_ZIGZAG, 2)
```

Comment:
I think the order of these should be flipped?
For example, it's been a common pattern to specify the upper range of capacitance at infinity - and I think as written, it wouldn't be able to find a package because it would look at the upper range?

However, if you choose a preferred number first, that would bound the range before the package selection logic runs.